### PR TITLE
[#6315] Add FAQ tooltip

### LINKED ISSUE #6315: Add support for configurable bottom-sheet tool tips
- [x] Implement bottom-sheet tool tip following the design in the front end  https://github.com/open-formulieren/formio-renderer/pull/334
- [x] Implement tool tip per field following the design in the front end https://github.com/open-formulieren/formio-renderer/pull/335
  - [x] Update field component schemas to include newly configurable tool tip related fields https://github.com/open-formulieren/types/pull/95
- [x] Allow multiple tool tips to be configured per field through the admin form builder https://github.com/open-formulieren/formio-builder/pull/300

**Additional notes for the form admin**
-  Allow bullet points & bold styles for WYSIWYG editor
- Configurable fields for the tool tips:
  - Title
  - Content

### PR DESCRIPTION
**Pull request by SonnyBA**
Closes the following bullet point, [from this ticket](https://github.com/open-formulieren/open-forms/issues/6315#issue-4523952977):

> Implement tool tip per field following the design in the front end

This branch was branched off from `open-formulieren/feature/6315-modal-sheet` as it depends on the updated Modal component.

TODO:

- [x] Update the pinned types version (to use the `FAQItem` types)

### CODE REVIEW

File: src/components/forms/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -86,6 +94,10 @@ const Checkbox: React.FC<CheckboxProps> = ({
   const descriptionId = description && !descriptionAsHelpText ? `${id}-description` : undefined;
 
   const ariaDescribedBy = [errorMessageId, descriptionId].filter(Boolean).join(' ');
+  const faqElements = faqItems.map((faqItem, index) => (
+    <FAQTooltip key={index} faqItem={faqItem} />
+  ));
```
Comment (sergei-maertens) [inline src/components/forms/Checkbox/Checkbox.tsx]: this map is repeated everywhere - wouldn't it make more sense to have a simple component that you call like this?

```tsx
<FAQItems items={faqItems} />
```
The map/loop itself would then be implement in the `FAQItems` component

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/components/forms/Checkbox/Checkbox.stories.ts [hunk 2]
```diff
@@ -52,6 +52,35 @@ export const WithTooltip: Story = {
...
+      {
+        label: 'Is this field applicable to me?',
+        content: 'This field is applicable if you are XYZ.',
+      },
+    ],
+  },
+  play: async ({canvasElement}) => {
+    const canvas = within(canvasElement);
+    const faqLabels1 = canvas.getAllByText('How do I fill in this field?');
+    const faqLabels2 = canvas.getAllByText('Is this field applicable to me?');
```
Comment (sergei-maertens) [inline src/components/forms/Checkbox/Checkbox.stories.ts]: queries `byRole` are better as they ensure that you have the accessibility requirements right - in this case, you'd do:

```tsx
canvas.getByRole('button', {name: 'How do I fill in this field?'});
```

File: src/components/forms/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -11,6 +12,7 @@ import Tooltip from '@/components/forms/Tooltip';
 import ValidationErrors from '@/components/forms/ValidationErrors';
 import {useFieldConfig} from '@/hooks';
 
+import FAQItems from '../FAQItems';
```
Comment (sergei-maertens) [inline src/components/forms/Checkbox/Checkbox.tsx]:
```suggestion
import FAQItems from '@/components/forms/FAQItems';
```
please use absolute imports - typically we use relative imports only when in the same 'package`, so only `./Foo` would be allowed.

File: src/components/forms/DateField/DateField.tsx [hunk 4]
```diff
@@ -158,6 +166,7 @@ const DateField: React.FC<DateFieldProps> = ({
     <FormField type="text" invalid={isInvalid} className="utrecht-form-field--openforms">
       {dateInput}
       <HelpText>{description}</HelpText>
+      <FAQItems items={faqItems} />
```
Comment (sergei-maertens) [inline src/components/forms/DateField/DateField.tsx]: I think you want validation errors to go above faqitems - if you have 5 items, it would push the errors off the screen.

File: package-lock.json [hunk 5]
```diff
@@ -2346,9 +2347,6 @@
         "arm64"
       ],
       "dev": true,
-      "libc": [
-        "glibc"
-      ],
```
Comment (sergei-maertens) [inline package-lock.json:2351]: where is this coming from? Are you perhaps using a different node version than what's in `.nvmrc`?

File: package.json [hunk 6]
```diff
@@ -77,6 +77,7 @@
     "@maykinmedia/eslint-config": "^2.0.0",
     "@open-formulieren/design-tokens": "^0.66.0",
     "@open-formulieren/leaflet-tools": "^1.0.0",
+    "@open-formulieren/types": "^1.4.1",
```
Comment (sergei-maertens) [inline package.json]: why was this moved into the dev dependencies?

File: src/components/forms/FAQItems.tsx
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx:1]: I would keep all the code related to the FAQItems together, ultimately that's the "public" API for the component that you expose to other components. `FAQTooltip` can be moved into this file, and doesn't need to be exported then.

File: src/components/forms/FAQItems.tsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
+
+import FAQTooltip from './FAQTooltip';
+
+export interface CheckboxProps {
+  /**
+   * FAQ tooltips to provide additional information that is not crucial but may
+   * assist users in filling out the field correctly.
+   */
+  items?: FAQItem[];
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]:
```suggestion
  items: FAQItem[];
```
be strict if possible - this component makes no sense if there are no item definitions - the call site can and should opt to pass a empty array as default or conditionally render: `faqItems && <FAQItems items={faqItems} />`.

That leads to a simpler implementation and less edge cases to test for this component.

File: src/components/forms/FAQItems.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
+
+import FAQTooltip from './FAQTooltip';
+
+export interface CheckboxProps {
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]: `CheckboxProps`?

File: src/components/forms/FAQItems.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]: I'd update the types to have this exported from the entrypoint: https://github.com/open-formulieren/types/blob/main/src/index.ts


```suggestion
import type {FAQItem} from '@open-formulieren/types';
```

File: src/components/forms/FAQItems.tsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+export interface CheckboxProps {
+  /**
+   * FAQ tooltips to provide additional information that is not crucial but may
+   * assist users in filling out the field correctly.
+   */
+  items?: FAQItem[];
+}
+
+const FAQItems: React.FC<CheckboxProps> = ({items = []}) => {
+  return items.map((item, index) => <FAQTooltip key={index} faqItem={item} />);
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]:
```suggestion
  return items.map(({label, content}, index) => <FAQTooltip key={index} label={label} content={content} />);
```
this is much more idiomatic react - you're passing primitives as props instead of an object, which works better with potential future memoization and is much more explicit + more ergonomic when using this component in isolation.

In its current form, you have some excessive nesting and if the `FAQItem` type gets extended with stuff that's not directly relevant for these components, you're dealing with friction and unrelated changes when another PR bumps the types version (you'll need to update stories and call sites).

File: src/components/forms/FAQTooltip.tsx [hunk 11]
```diff
@@ -0,0 +1,52 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
+import {HTMLContent, Icon as UtrechtIcon} from '@utrecht/component-library-react';
+import DOMPurify from 'dompurify';
+import type {KeyboardEvent} from 'react';
+import {useState} from 'react';
+
+import Icon from '@/components/icons';
+
+import Modal from '../modal';
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]: please avoid relative imports to parent directories, it's so hard to see where the component is being imported from when reviewing a PR

File: src/components/forms/FAQTooltip.tsx [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import './FAQTooltip.scss';
+
+export interface FAQTooltipProps {
+  faqItem: FAQItem;
+}
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]: as mentioned above - let's split this out and make it explicit. It's Typescript's job to make sure that our FAQItem at call sites matches the expected props :)

File: src/components/forms/FAQTooltip.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+const FAQTooltip: React.FC<FAQTooltipProps> = ({faqItem}) => {
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]: Given the DOMPurify sanitize happening on every render (and the parent rerender triggering a re-render of all children), this is potentially expensive. Combined with making the props primitives and atomic, I would immediately opt for [`React.memo`](https://react.dev/reference/react/memo) to avoid performance issues in large forms with many faq items and/or complex content markup.

File: src/components/forms/FAQTooltip.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  const [isOpen, setIsOpen] = useState(false);
+
+  const sanitizedContent = DOMPurify.sanitize(faqItem.content, {USE_PROFILES: {html: true}});
+  const modalSheet = (
+    <Modal
+      title={faqItem.label}
+      variant={'sheet'}
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]:
```suggestion
      variant="sheet"
```

File: src/components/forms/FAQTooltip.tsx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  const dialogOpenKeys = ['Enter', 'Space'];
+  return (
+    <>
+      <div
+        role="button"
+        tabIndex={0}
+        className="openforms-faq-tooltip"
+        onClick={() => setIsOpen(true)}
+        onKeyDown={(event: KeyboardEvent) => dialogOpenKeys.includes(event.code) && setIsOpen(true)}
+      >
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]: This really should be a proper `<button>` element instead of a div pretending to a button. Also, we really should use existing NL DS components to make this themeable.

I would however opt for a custom modifier so that custom themes can apply CSS overrides if necessary: a final  classname of `utrecht-link-button openforms-faq-item` should get us there for all the styling needs.

File: src/components/forms/FAQTooltip.tsx [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        <UtrechtIcon>
+          <Icon icon="tooltip" className="openforms-faq-tooltip__icon" />
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]:
```suggestion
          <Icon icon="tooltip" />
```
I don't think the additional class name is necessary

File: src/components/forms/FAQTooltip.tsx [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        </UtrechtIcon>
+
+        <span className="openforms-faq-tooltip__title">{faqItem.label}</span>
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.tsx]:
```suggestion
        <span className="openforms-faq-item__title">{faqItem.label}</span>
```
good to keep it so that we can provide support for custom design tokens

File: src/components/forms/FAQItem.scss [hunk 18]
```diff
@@ -0,0 +1,17 @@
+@use '@utrecht/icon-css';
+
+@use '@/scss/bem';
+
+.openforms-faq-tooltip {
+  margin-block-start: var(--of-faq-tooltip-margin-block, 8px);
```
Comment (sergei-maertens) [inline src/components/forms/FAQItem.scss]: blocks/components are *not* allowed to set margins, because that affects the world outside of themselves

I'd to the route of a container element:

```scss
.openforms-faq-items {
  display: flex;
  flex-direction: column;
  row-gap: var(--of-faq-items-row-gap, 8px);
  padding-block-start: var(--of-faq-items-padding-block-start, 8px); // not sure if this is the best approach
}
```

This sets some default spacing, but allows theme designers to use different spacing values if they desire so.

File: src/components/forms/FAQTooltip.scss [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+  padding-block-end: var(--of-faq-tooltip-padding-block, 11px);
+  padding-block-start: var(--of-faq-tooltip-padding-block, 11px);
+  padding-inline-end: var(--of-faq-tooltip-padding-inline, 16px);
+  padding-inline-start: calc(
+    var(--of-faq-tooltip-padding-inline, 16px) - var(--_border-left-width)
+  );
+
+  &__title:before {
+    content: ' ';
+  }
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.scss]: All this should not be necessary when you use the `.utrecht-link-button` component, because it already defines a column-gap token that creates the spacing between icon and label. That's also the right approach:

* it ensures accessibility
* it provides consistent spacing for button/link-like elements
* it can be themed through design tokens

File: src/components/forms/FAQTooltip.scss [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+.openforms-faq-tooltip {
+  margin-block-start: var(--of-faq-tooltip-margin-block, 8px);
+  padding-block-end: var(--of-faq-tooltip-padding-block, 11px);
+  padding-block-start: var(--of-faq-tooltip-padding-block, 11px);
+  padding-inline-end: var(--of-faq-tooltip-padding-inline, 16px);
+  padding-inline-start: calc(
+    var(--of-faq-tooltip-padding-inline, 16px) - var(--_border-left-width)
+  );
+
+  &__title:before {
```
Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.scss]: also, you should really be using the `@include bem.element('title')` sass selector here. You can then still include the pseudo element:

```scss
.openforms-faq-item {
  @include bem.element('title') {
    &:before { ... }
  }
}
```

Comment (sergei-maertens) [inline src/components/forms/FAQTooltip.scss]: I think we want to be prepared for some potential different typography through opt-in design tokens. If they're not defined, they'll just fall back to the defaults of the utrecht-link-button component (so use the link tokens), I believe:

```scss
.openforms-faq-item {
  @include bem.element('title') {
    color: var(--of-faq-item-title-color);
    font-size: var(--of-faq-item-title-font-size);
    font-weight: var(--of-faq-item-title-font-weight);
    line-height: var(--of-faq-item-title-line-height);
    text-decoration: var(--of-faq-item-title-text-decoration);
    
    &:hover {
      text-decoration: var(--of-faq-item-title-hover-text-decoration);
    }
  }
}
```
I don't think we need to define any of those design tokens ourselves at this point.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
Please also improve you commit hygiene:

* The types version bump should be it's own atomic commit
* The implementation of the `FAQItems` component should be its own atomic feature commit
* Adding the feature to all components is its own feature commit too

I'm also missing more information in the commit message:

* why the types dependency needed bumping, and why it was moved from dependencies to dev dependencies
* why you opted for `div`s with `role="button"` instead of proper button elements
* why there are multiple text nodes for the same label of an FAQ item

File: src/components/forms/FAQItems.tsx (hunk 9 above)
Comment (SonnyBA) [reply src/components/forms/FAQItems.tsx]: See https://github.com/open-formulieren/types/pull/99 for the change in the types repository.

File: src/components/forms/Checkbox/Checkbox.tsx [hunk 21]
```diff
@@ -1,10 +1,12 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
```
Comment (sergei-maertens) [inline src/components/forms/Checkbox/Checkbox.tsx]: TODO: once new version of types is available, update the import

File: src/components/forms/DateField/DateField.tsx [hunk 22]
```diff
@@ -1,7 +1,9 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
```
Comment (sergei-maertens) [inline src/components/forms/DateField/DateField.tsx]: TODO: once new version of types is available, update the import

File: src/components/forms/DateTimeField/DateTimeField.tsx [hunk 23]
```diff
@@ -1,3 +1,4 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
```
Comment (sergei-maertens) [inline src/components/forms/DateTimeField/DateTimeField.tsx]: TODO: once new version of types is available, update the import

File: src/components/forms/EditGrid/EditGrid.tsx [hunk 24]
```diff
@@ -1,3 +1,4 @@
+import type {FAQItem} from '@open-formulieren/types/dist/common';
```
Comment (sergei-maertens) [inline src/components/forms/EditGrid/EditGrid.tsx]: TODO: once new version of types is available, update the import

File: src/components/forms/InputGroup/InputGroup.tsx
Comment (sergei-maertens) [inline src/components/forms/InputGroup/InputGroup.tsx:1]: Do we need the support for FAQ items in this component? I don't think so.

The problem here is that there's no slot for the validation errors, meaning validation errors going below the input group appear after the FAQ Items, which is opposite of what we want.

I think this component should not support the prop, and the components that use `InputGroup` are responsible of rendering the faq items themselves, after the validation errors display.

File: src/components/forms/FAQItems.tsx [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+  );
+
+  const dialogOpenKeys = ['Enter', 'Space'];
+  return (
+    <>
+      <LinkButton
+        tabIndex={0}
+        className="openforms-faq-items__item"
+        onClick={() => setIsOpen(true)}
+        onKeyDown={(event: KeyboardEvent) => dialogOpenKeys.includes(event.code) && setIsOpen(true)}
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]: does this not happen automatically through the `onClick`? I'd expect the native button element to cover that, once it has focus.

File: src/components/forms/FAQItems.tsx [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+    <Modal title={label} variant="sheet" closeModal={() => setIsOpen(false)} isOpen={isOpen}>
+      <HTMLContent dangerouslySetInnerHTML={{__html: sanitizedContent}} />
+    </Modal>
+  );
+
+  const dialogOpenKeys = ['Enter', 'Space'];
+  return (
+    <>
+      <LinkButton
+        tabIndex={0}
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]: When I remove this in the DOM in storybook, I can still tab properly, so I don't think this is necessary (`<button>` elements can be tabbed by default).

File: src/components/forms/FAQItems.tsx [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+        className="openforms-faq-items__item"
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]:
```suggestion
        className="openforms-faq-item"
```
the plural form `openforms-faq-items` is for the container, the singular form for a single item

File: src/components/forms/FAQItems.tsx [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+        onClick={() => setIsOpen(true)}
+        onKeyDown={(event: KeyboardEvent) => dialogOpenKeys.includes(event.code) && setIsOpen(true)}
+      >
+        <UtrechtIcon>
+          <Icon icon="tooltip" />
+        </UtrechtIcon>
+
+        <span className="openforms-faq-items__title">{label}</span>
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]:
```suggestion
        <span className="openforms-faq-item__title">{label}</span>
```
here too - use the singular form for proper BEM semantics

File: src/components/forms/FAQItems.tsx [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+      </LinkButton>
+      {modalSheet}
+    </>
+  );
+};
+
+const FAQItems: React.FC<FAQItemsProps> = ({items = []}) => {
```
Comment (sergei-maertens) [inline src/components/forms/FAQItems.tsx]:
```suggestion
const FAQItems: React.FC<FAQItemsProps> = ({items}) => {
```
default is not necssary, because the prop cannot be undefined/null from the types

File: src/components/forms/FAQItem.scss [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+    font-size: var(--of-faq-item-title-font-size);
+    font-weight: var(--of-faq-item-title-font-weight);
+    line-height: var(--of-faq-item-title-line-height);
+    text-decoration: var(--of-faq-item-title-text-decoration);
+
+    &:hover {
+      text-decoration: var(--of-faq-item-title-hover-text-decoration);
+    }
+  }
+}
```
Comment (sergei-maertens) [inline src/components/forms/FAQItem.scss:26]:
```suggestion
.openforms-faq-items {
  display: flex;
  flex-direction: column;
  row-gap: var(--of-faq-items-row-gap, 8px);
  align-items: flex-start;
  padding-block-start: var(--of-faq-items-padding-block-start, 8px);
}

.openforms-faq-item {
  @include bem.element('title') {
    color: var(--of-faq-item-title-color);
    font-size: var(--of-faq-item-title-font-size);
    font-weight: var(--of-faq-item-title-font-weight);
    line-height: var(--of-faq-item-title-line-height);
    text-decoration: var(--of-faq-item-title-text-decoration);

    &:hover {
      text-decoration: var(--of-faq-item-title-hover-text-decoration);
    }
  }
}
```

Note that in your version you're mixing design tokens - a component with name `openforms-faq-items` cannot have a design token `of-faq-item` (singular) - the component names don't match!

The proper approach here is to just split in two components/blocks - one for the container, and one for an individual item.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/components/forms/FAQItems.tsx (hunk 25 above)
Comment (SonnyBA) [reply src/components/forms/FAQItems.tsx]: It does :exploding_head: , TIL.

**Review by sergei-maertens — APPROVED** [review]
